### ISSUE TITLE
FR(monitoring): (Re)establish metrics/monitoring for former galley/citadel functionality in istiod

### ISSUE THREAD

**Issue by douglas-reid** [istio/istio#22508]
**Describe the feature request**

In the move to `istiod`, the code paths for the functionality formerly provided by `istio-galley` and `istio-citadel` do not export the metrics that were previously used for monitoring. While we don't need 100% duplication, we should have proper monitoring of those code paths and exposed functionality.

**Describe alternatives you've considered**

[ X ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ X ] Networking
[ ] Performance and Scalability
[ X ] Policies and Telemetry
[ X ] Security
[ ] Test and Release
[ X ] User Experience
[ ] Developer Infrastructure

**Additional context**

**Comment by howardjohn** [comment]
https://github.com/istio/istio/pull/22507 gets us a bit of this

**Comment by myidpt** [comment]
Thanks for creating this!

**Comment by JimmyCYJ** [comment]
There is another one about exporting Citadel metrics to grafana https://github.com/istio/istio/issues/21004

**Comment by brian-avery** [comment]
@lei-tang Can you please update this?

**Comment by lei-tang** [comment]
I am working on other items and have not been able to work on this issue.

**Comment by howardjohn** [comment]
@douglas-reid do we have any info on specific metrics we are interested in?

**Comment by douglas-reid** [comment]
@howardjohn we used to track number of virtual services and destination rules (with dimensions for namespace, etc.), among other items. i think that code still exists, just isn't exercised. It'd be nice to have config-oriented metrics. @gargnupur was the expert on these metrics.

**Comment by howardjohn** [comment]
Got it. We currently have information about *events* for these but the request is to add a gauge of current number basically?

api-server seems to exposes this btw, similar query: `sum(etcd_object_counts{resource=~\"pods|namespaces|services|gateways.networking.istio.io|virtualservices.networking.istio.io\"}) by (resource)`. 

Do we just want CRDs or also Service, Endpoint, etc?

I am not against adding it if there is a use case for it, just want to make sure there is a real use case, beyond just galley did it previously

**Comment by douglas-reid** [comment]
@howardjohn i guess we could grab the API server version of them as a placeholder (the trick will be using them in non-prom cases). i don't feel strongly about Service and Endpoint, but i feel like it is useful for Istio configuration.

We had previously added gauges for some of these Istio CRDs to the mesh dashboard, and had talked about using the change events to annotate service dashboards, etc. (to help with correlating behavioral changes with config changes).

They were originally meant to be the Istio analogue to Kube State Metrics. We had a separate repo for awhile with an example image that would run and generate the metrics as a demo. It used the same code generation bits that later became istio/client-go.

We added them to galley because it was felt that we wanted to monitor config as it was processed by various bits of Istio (how many exist, but failed validation, etc.). The plan was to add the monitoring at each stage (galley, pilot/mixer). I think we only got to galley.

We could go back to having it live somewhere like istio/ecosystem and be based on istio/client-go and be opt-in, I guess. But then we should definitely clean-up our dashboards.
